src/drivers/gpio/gpio.h

```diff
@@ -0,0 +1,46 @@
+/**
+ * GPIO module
+ * Driver for GPIO pins
+ */
+#ifndef GPIO_H
+#define GPIO_H
+
+#include "common.h"
+
+/**
+ * Enum representing GPIO pin level
+ */
+typedef enum {
+    GPIO_LEVEL_LOW,
+    GPIO_LEVEL_HIGH
+} gpio_level_t;
+
+/**
+ * Enum representing a connected GPIO pin
+ */
+typedef enum {
+    GPIO_PIN_RED_LED,
+    GPIO_PIN_GREEN_LED,
+    GPIO_PIN_BLUE_LED,
+    GPIO_PIN_COUNT // Enum max value
+} gpio_pin_t;
+
+/**
+ * Read the current level of `pin` into `level`. Non-blocking. Not thread-safe
+ * if multiple threads interact with the same pin.
+ */
+w_status_t gpio_read(gpio_pin_t pin, gpio_level_t *level);
```

Comment:
Does the hal have a fail condition here? if not for simplicity it might be simpler to just return the value. If that is the case pls put the output first

Reply:
hal doesnt have a fail but its possible that the `pin` input is invalid

Reply:
or if the cmd required any further processing in the future (for example we decide to add mutex), it would have even more fail conditions. Then we'd have to go back and change this to return status and everything using it.

Yeah that's unlikely but returning status has no downsides only future-proofing benefits imo